### PR TITLE
host and storage checks

### PR DESCRIPTION
**Pull request by DenisMishin**

### CODE REVIEW

File: monitoring/host.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/dustin/go-humanize"
+)
+
+const (
+	hostCheckerID = "cpu-ram"
+)
+
+// HostChecker validates CPU and RAM requirements
+type HostChecker struct {
+	// MinCPU minimum amount of logical CPUs
```
Comment (r0mant) [inline monitoring/host.go]: "MinCPU is minimum required amount of..."

File: monitoring/host.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	MinCPU int
+	// MinRamBytes minimum amount of RAM
```
Comment (r0mant) [inline monitoring/host.go]: "MinRAMBytes is minimum required amount of..."

File: monitoring/host.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	MinRAMBytes uint64
+}
+
+func (c *HostChecker) Name() string {
```
Comment (r0mant) [inline monitoring/host.go]: Godoc.

File: monitoring/host.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	return hostCheckerID
+}
+
+func (c *HostChecker) Check(ctx context.Context, reporter health.Reporter) {
```
Comment (r0mant) [inline monitoring/host.go]: Godoc.

File: monitoring/os.go [hunk 5]
```diff
@@ -79,7 +79,7 @@ func (c *OSChecker) Check(ctx context.Context, reporter health.Reporter) {
 
 	reporter.Add(&pb.Probe{
 		Checker: osCheckerID,
-		Detail:  fmt.Sprintf("%s %s is not supported, please see https://gravitational.com/docs/pack/#distributions", id, versionID),
```
Comment (r0mant) [inline monitoring/os.go]: Why did you remove this hint? It seemed useful.

File: monitoring/storage.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+	"github.com/gravitational/satellite/utils"
+	"github.com/gravitational/trace"
```
Comment (r0mant) [inline monitoring/storage.go]: This import belongs to the 3rd group.

File: monitoring/storage.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+	cycles                = 1024
+	stRdonly              = int64(1)
+)
+
+// Name returns name of the checker
+func (c *StorageChecker) Name() string {
+	return fmt.Sprintf("%s(%s)", storageWriteCheckerID, c.Path)
+}
+
+func (c *StorageChecker) Check(ctx context.Context, reporter health.Reporter) {
```
Comment (r0mant) [inline monitoring/storage.go]: Godoc.

File: monitoring/storage.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+		c.checkWriteSpeed(ctx, reporter))
+}
+
+// cleanPath returns fully evaluated path with symlinks resolved
+// if path doesn't exist but should be created, then it returns
+// first available parent directory, and checks will be applied to it
+func (c *StorageChecker) evalPath() error {
+	fi, err := os.Stat(c.Path)
+
+	if os.IsNotExist(err) && !c.WillBeCreated {
```
Comment (r0mant) [inline monitoring/storage.go]: What if other type of error returned? It seems like it's ignored.

File: monitoring/storage.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+	p := c.Path
+	for {
+		fi, err = os.Stat(p)
+		if err == nil && fi.IsDir() {
+			c.path = p
+			return nil
+		} else if err == nil && !fi.IsDir() {
+			return trace.BadParameter("%s is not a directory", p)
+		}
```
Comment (r0mant) [inline monitoring/storage.go]: Non-nil err is ignored.

File: monitoring/storage.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+		return trace.Wrap(err)
+	}
+
+	probe := &pb.Probe{Checker: c.Name()}
+
+	if utils.StringInSlice(c.Filesystems, mnt.SysTypeName) {
+		probe.Status = pb.Probe_Running
+	} else {
+		probe.Status = pb.Probe_Failed
+		probe.Detail = fmt.Sprintf("path %s requires %v, belongs to %s mount point of type %s",
```
Comment (r0mant) [inline monitoring/storage.go]: I would add "requires _filesystem_" for clarity.

File: monitoring/storage.go [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+	var file *os.File
+	file, err = ioutil.TempFile(c.path, "probe")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	defer func() {
+		err_c := file.Close()
+		err_r := os.Remove(file.Name())
+		err = trace.NewAggregate(err_c, err_r)
```
Comment (r0mant) [inline monitoring/storage.go]: I would do:

```go
defer func() {
  err = trace.NewAggregate(
    trace.ConvertSystemError(file.Close()),
    trace.ConvertSystemError(os.Remove(file.Name()))
}
```

File: monitoring/storage.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+	err = file.Sync()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	elapsed := time.Since(start).Seconds()
+	bps := uint64(blockSize * cycles / elapsed)
+
+	if bps >= c.MinBytesPerSecond {
+		return nil
```
Comment (r0mant) [inline monitoring/storage.go]: Don't we need to add a successful probe in this case to indicate the check passed? I think we do that in other places?

File: monitoring/storage.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	if avail < c.MinFreeBytes {
+		reporter.Add(&pb.Probe{
+			Checker: c.Name(),
+			Detail: fmt.Sprintf("%s available space left on %s, minimum of %s required",
+				humanize.Bytes(avail), c.Path, humanize.Bytes(c.MinFreeBytes)),
+			Status: pb.Probe_Failed,
+		})
+	}
+
+	return nil
```
Comment (r0mant) [inline monitoring/storage.go]: Same as below - maybe add a successful probe to indicate the check passed like we do for other checks?

File: monitoring/os.go (hunk 5 above)
Comment (DenisMishin) [reply monitoring/os.go]: because it's used in manifest checks, and i.e. Anypoint manifest won't contain ubuntu as supported OS, and the error message will be misleading as it'll refer end-user to Gravitational docs where we will claim Ubuntu is supported.

File: monitoring/storage.go (hunk 9 above)
Comment (DenisMishin) [reply monitoring/storage.go]: yes, deliberately; we just go one level up.

File: monitoring/storage.go (hunk 12 above)
Comment (DenisMishin) [reply monitoring/storage.go]: it's there, see line 75

File: monitoring/storage.go (hunk 13 above)
Comment (DenisMishin) [reply monitoring/storage.go]: it's there, see line 75

File: monitoring/host.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+// Check runs check
+func (c *HostChecker) Check(ctx context.Context, reporter health.Reporter) {
+	failed := false
+
+	var mem sigar.Mem
+	err := mem.Get()
+	if err != nil {
+		reporter.Add(NewProbeFromErr(hostCheckerID, "querying memory info", trace.Wrap(err)))
```
Comment (a-palchikov) [inline monitoring/host.go]: Why is this not an error for this checker (as in `failed = true`)? Then, you could do away w/ `failed` and just use `len(reporter.GetProbes()) != 0` instead

File: monitoring/storage.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+// StorageChecker verifies volume matches requirements
+type StorageChecker struct {
+	// Path represents volume to be checked
+	Path string
+	// WillBeCreated when true, then all checks will be applied to first existing dir, or fail otherwise
+	WillBeCreated bool
+	// path is normalized path, representing parent directory
+	// in case Path does not exist yet
+	path string
```
Comment (a-palchikov) [inline monitoring/storage.go]: This should've slided to after the exported attributes

File: monitoring/storage.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+	p := c.Path
+	for {
+		fi, err := os.Stat(p)
+
+		if err != nil && !os.IsNotExist(err) {
+			return trace.ConvertSystemError(err)
+		}
+
+		if os.IsNotExist(err) && !c.WillBeCreated {
+			return trace.BadParameter("%s does not exist", c.Path)
```
Comment (a-palchikov) [inline monitoring/storage.go]: `NotFound`

File: monitoring/storage.go [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.NewAggregate(c.checkFsType(ctx, reporter),
+		c.checkCapacity(ctx, reporter),
+		c.checkWriteSpeed(ctx, reporter))
+}
+
+// cleanPath returns fully evaluated path with symlinks resolved
```
Comment (a-palchikov) [inline monitoring/storage.go]: s/cleanPath/evalPath

File: monitoring/storage.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+
+	sort.Sort(childPathFirst(mounts.List))
+
+	for _, mnt := range mounts.List {
+		if strings.HasPrefix(cleanpath, mnt.DirName) {
+			return &mnt, nil
+		}
+	}
+
+	return nil, trace.BadParameter("failed to locate filesystem for %s", path)
```
Comment (a-palchikov) [inline monitoring/storage.go]: `NotFound`

File: monitoring/storage_test.go [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+	require.NoError(t, err)
+
+	t.Logf("%+v", tmp)
+
+	StorageChecker{
+		Path: "/tmp",
+	}.probe(t, "no conditions", shallSucceed)
+
+	StorageChecker{
+		Path:              path.Join("/tmp", "do-not-exist"),
```
Comment (a-palchikov) [inline monitoring/storage_test.go]: `does-not-exist`
